Back-To-School/Assets/BackToSchool/Scripts/GameManagement/Game.cs

```diff
@@ -0,0 +1,179 @@
+        [SerializeField] private GameOverPresenter _gameOverPresenter;
+        [SerializeField] private PausePresenter _pausePresenter;
+
+        [SerializeField] private PlayerController _playerPrefab;
+        [SerializeField] private GameObject _inputManagerPrefab;
+        [SerializeField] private GameObject _enemySpawnerPrefab;
+        [SerializeField] private Camera _mainCamera;
+
+        private EnemySpawner _enemySpawner;
+        private Action _saveAction;
+        private GameManager _gameManager;
+        private IInputManager _inputManager;
+        private LevelSystem _levelSystem;
+        private PlayerController _player;
+        private PlayerInputProvider _playerInput;
+        private PlayerStats _playerStats;
+        private StatsManager _statsManager;
+        private SaveSystem _saveSystem;
+
+        private float _gameOverDelay = 1f;
+        private bool _isGamePaused;
+        private bool _isPlayerDead;
+
+        public void Initialize(SaveSystem saveSystem, GameManager gameManager, GameParameters parameters)
+        {
+            _saveSystem   = saveSystem;
+            _gameManager  = gameManager;
+            _playerStats  = new PlayerStats();
+            _levelSystem  = new LevelSystem();
+            _statsManager = new StatsManager();
+
+            _player       = Instantiate(_playerPrefab, transform.position, Quaternion.identity);
+            _inputManager = Instantiate(_inputManagerPrefab, transform.position, Quaternion.identity).GetComponent<InputManager>();
+            _enemySpawner = Instantiate(_enemySpawnerPrefab, transform.position, Quaternion.identity).GetComponent<EnemySpawner>();
+
+            SubscribeEvents();
+            _statsManager.Initialize(_playerStats, parameters.InitialLevel);
+
+            _mainCamera.GetComponent<CameraFollow>().SetTarget(_player.transform);
+            _enemySpawner.SetTarget(_player.gameObject);
+
+            _playerInput = new PlayerInputProvider(_mainCamera);
+            _inputManager.Subscribe(_playerInput);
+            _player.Initialize(_playerInput, _playerStats);
+
+            var pauseInput = new PauseInputProvider();
+            _inputManager.Subscribe(pauseInput);
+            pauseInput.Cancelled += OnGameStopped;
+
+            if (parameters.IsNewGame)
+            {
+                _saveSystem.ResetPlayerProgress();
+                _player.InitializeHealth();
+                _player.WeaponController.InitializeWeapon();
+            }
+            else
+                _saveSystem.LoadPlayerProgress(_player, _levelSystem);
+        }
+
+        private void SubscribeEvents()
+        {
+            _gameOverPresenter.Restarted += RestartGame;
+            _pausePresenter.Restarted    += RestartGame;
+            _pausePresenter.Continued    += ContinueGame;
+            _pausePresenter.MenuReturned += ReturnToMenu;
+
+            _player.WeaponController.AmmoChanged    += _hudPresenter.OnAmmoChanged;
```

Comment:
And I'm not sure that it's a good idea to make WeaponController accissible for others. I understand that this may require to make a lot of spare code, but will be a bit more flexible. Moreover, interfaces would be cleaner and less coherent